File: ObjC/SampleApp/InterstitialViewController.m
///
/// @file InterstitialViewController.m
/// @brief Implementation for InterstitialViewController
///
/// @copyright Copyright Yahoo, Licensed under the terms of the Apache 2.0 license . See LICENSE file in project root for terms.
///

#import "InterstitialViewController.h"
#import "AdPlacementConsts.h"
#import <YahooAds/YahooAds.h>

@interface InterstitialViewController () <YASInterstitialAdDelegate>

@property (nonatomic, strong) IBOutlet UIButton *requestButton;
@property (nonatomic, strong) IBOutlet UIButton *showButton;
@property (nonatomic, strong) YASInterstitialAd *interstitialAd;

@end

@implementation InterstitialViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [self.showButton setEnabled:NO];
}

- (IBAction)loadInterstitial:(id)sender
{
    [self.showButton setEnabled:NO];
    self.interstitialAd = [[YASInterstitialAd alloc] initWithPlacementId:kSampleAppInterstitialAdPlacementID];
    self.interstitialAd.delegate = self;
    [self.interstitialAd loadWithPlacementConfig:nil];
}

- (IBAction)displayInterstitial:(id)sender
{
    [self.interstitialAd setImmersiveEnabled:YES];
    [self.interstitialAd setEnterAnimationId:UIModalTransitionStyleCrossDissolve];
    [self.interstitialAd setExitAnimationId:UIModalTransitionStyleCrossDissolve];
    [self.interstitialAd showFromViewController:self];
}

#pragma mark - YASInterstitialAdDelegate

- (void)interstitialAdDidLoad:(YASInterstitialAd *)interstitialAd
{
    NSLog(@"Interstitial ad did load");
    self.interstitialAd = interstitialAd;
    
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        __strong typeof(weakSelf) strongSelf = weakSelf;
        if (strongSelf == nil) {
            return;
        }
        strongSelf.showButton.enabled = YES;
    });
}

- (void)interstitialAdLoadDidFail:(YASInterstitialAd *)interstitialAd withError:(YASErrorInfo *)errorInfo
{
    NSLog(@"Interstitial ad load failed with error: %@", errorInfo);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.showButton setEnabled:NO];
    });
}

- (void)interstitialAdDidFail:(YASInterstitialAd *)interstitialAd withError:(YASErrorInfo *)errorInfo
{
    NSLog(@"Interstitial ad failed to display with error: %@", errorInfo);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.showButton setEnabled:NO];
    });
}

- (void)interstitialAdDidShow:(YASInterstitialAd *)interstitialAd
{
    NSLog(@"Interstitial ad was shown");
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.showButton setEnabled:NO];
    });
}

- (void)interstitialAdDidClose:(YASInterstitialAd *)interstitialAd
{
    NSLog(@"Interstitial ad was closed");
}

- (void)interstitialAdClicked:(YASInterstitialAd *)interstitialAd
{
    NSLog(@"Interstitial ad was clicked");
    [YahooAudiences logEventWithEventType:YahooAudiencesEventAdClick andParamBuilder:nil];
}

- (void)interstitialAdDidLeaveApplication:(YASInterstitialAd *)interstitialAd
{
    NSLog(@"Interstitial ad did cause user to leave application");
}

- (void)interstitialAdEvent:(YASInterstitialAd *)interstitialAd source:(NSString *)source eventId:(NSString *)eventId arguments:(NSDictionary<NSString *, id> *)arguments
{
    NSLog(@"Interstitial ad event occurred with ID: %@", eventId);
}

@end

File: Swift/SampleApp/InterstitialViewController.swift
///
/// @file InterstitialViewController.swift
/// @brief Implementation for InterstitialViewController
///
/// @copyright Copyright Yahoo, Licensed under the terms of the Apache 2.0 license . See LICENSE file in project root for terms.
///

import Foundation
import UIKit
import YahooAds

class InterstitialViewController : UIViewController {

    // MARK: - Properties
    
    @IBOutlet weak var requestAdButton: UIButton?
    @IBOutlet weak var showButton: UIButton?

    private var ad: YASInterstitialAd?

    override func viewDidLoad() {
        super.viewDidLoad()
        loadInterstitial(nil)
    }
    
    // MARK: - User Interaction
    
    @IBAction private func loadInterstitial(_ sender: AnyObject?) {
        showButton?.isEnabled = false
        requestAdButton?.isEnabled = false
        
        ad = YASInterstitialAd(placementId: AdPlacementData.interstitialId)
        ad?.delegate = self
        ad?.load(with: nil)
    }

    @IBAction private func displayInterstitial(_ sender: AnyObject) {
        ad?.isImmersiveEnabled = true
        ad?.enterAnimationId = UIModalTransitionStyle.crossDissolve.rawValue
        ad?.exitAnimationId = UIModalTransitionStyle.crossDissolve.rawValue
        ad?.show(from: self)
    }
}

// MARK: - YASInterstitialAdDelegate

extension InterstitialViewController: YASInterstitialAdDelegate {

    func interstitialAdDidLoad(_ interstitialAd: YASInterstitialAd) {
        print("Interstitial ad did load")
        ad = interstitialAd
        
        DispatchQueue.main.async {
            self.showButton?.isEnabled = true
            self.requestAdButton?.isEnabled = true
        }
    }
    
    func interstitialAdLoadDidFail(_ interstitialAd: YASInterstitialAd, withError errorInfo: YASErrorInfo) {
        print("Interstitial ad failed to display with error:  \(errorInfo)")
        ad = nil
        
        DispatchQueue.main.async {
            self.showButton?.isEnabled = false
            self.requestAdButton?.isEnabled = true
        }
    }
    
    func interstitialAdDidFail(_ interstitialAd: YASInterstitialAd, withError errorInfo: YASErrorInfo) {
        print("Interstitial ad failed with error:  \(errorInfo)")
    }
    
    func interstitialAdDidShow(_ interstitialAd: YASInterstitialAd) {
        print("Interstitial ad was shown")
        
        DispatchQueue.main.async {
            self.showButton?.isEnabled = false
            self.requestAdButton?.isEnabled = true
        }
    }
    
    func interstitialAdDidClose(_ interstitialAd: YASInterstitialAd) {
        print("Interstitial ad was closed")
    }
    
    func interstitialAdClicked(_ interstitialAd: YASInterstitialAd) {
        print("Interstitial ad was clicked")

        YahooAudiences.logEvent(YahooAudiencesEvent.adClick, with: nil)
    }
    
    func interstitialAdDidLeaveApplication(_ interstitialAd: YASInterstitialAd) {
        print("Interstitial ad did cause user to leave application")
    }
    
    func interstitialAdEvent(_ interstitialAd: YASInterstitialAd, source: String, eventId: String, arguments: Dictionary<String, Any>?) {
        print("Interstitial ad event occurred with ID: \(eventId)")
    }
}
